### PR TITLE
LV-End-Point

### LINKED ISSUE #6: [WebApi] GetAll products end point
Add a query to retrieve all products in the [WebAPI] project

return type should be PegedResult<T> (keep in mind that this one is generic type)
{
    List<T> Items
    int Page
    byte PageSize
}

### PR DESCRIPTION
**Pull request by LVarsaminov**
# Summary
[Basic information about the purpose of the PR]
Have completed and tested the End-Point Issue.

# Technical details
[any technical changes or details that need highlighting]
Created generic class PagedResult.cs
Refactored IProductRepository and ProductRepository.cs.
Added GetAllProductsQuery and GetAllProductsQueryHandler
Added ProductDTO.cs
Added ProductExtensions.cs which maps the Product.cs properties to the ProductDTO's.
Created the Products.Controller and configured MediatR
Configured Swagger in SportsGoods.Api Program.cs 
Created tests to check the functionality of the QueryHandler

# Test
[how should this PR be tested and things to watch out for]
1. Start the SportsGoods.Api project.
2. Swagger UI should open
3. You should see GetAllProducts endpoint
4. When you click try it out you should be able to enter custom input on both of the parameters or use their default values
5.  If parameters are correct you should see a list of products with count of the pageSize parameter.
6. If one of the parameters is negative you will see an exception message
7. If you have tried to access a non-existing page (which is positive, so we don't cause the exception), you should see an empty list
8. Additionally if you want to run the tests they should all pass as 'Success'

### CODE REVIEW

File: SportsGoods.Api/Program.cs [hunk 1]
```diff
@@ -1,25 +1,50 @@
...
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.OpenApi.Models;
+using SportsGoods.App.Queries;
+using SportsGoods.App.QueryHandlers;
+using SportsGoods.Core.Interfaces;
+using SportsGoods.Data.DAL;
+using SportsGoods.Data.Repositories;
+using System.Reflection;
```
Comment (clangelov) [inline SportsGoods.Api/Program.cs]: Remove unused 'using'(s)

image

Ctrl+R+G in VS

File: SportsGoods.Api/SportsGoods.Api.csproj [hunk 2]
```diff
@@ -8,7 +8,9 @@
 
   <ItemGroup>
     <PackageReference Include="MediatR" Version="12.2.0" />
+    <PackageReference Include="MediatR.Extensions.Microsoft.DependencyInjectionFixed" Version="5.1.2" />
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.3" />
+    <PackageReference Include="Microsoft.ServiceFabric.Actors" Version="7.1.1951" />
```
Comment (clangelov) [inline SportsGoods.Api/SportsGoods.Api.csproj]: Remove unused packages

File: SportsGoods.App/QueryHandlers/GetAllProductsQueryHandler.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+            var pagedResult = new PagedResult<ProductDTO>
+            {
+                Items = productDtos,
+                Page = request.PageNumber,
+                PageSize = request.PageSize,
+                TotalCount = pagedProducts.TotalCount
+            };
+
+            return pagedResult;
```
Comment (clangelov) [inline SportsGoods.App/QueryHandlers/GetAllProductsQueryHandler.cs]: Reduce boilerplate code 

```
return new PagedResult<ProductDTO>
{
    Items = pagedProducts.Items.Select(x => x.ConvertToDto()).ToList(),
    Page = request.PageNumber,
    PageSize = request.PageSize,
    TotalCount = pagedProducts.TotalCount
};
```

File: SportsGoods.Core/Models/PagedResult.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace SportsGoods.Core.Models
+{
+    public class PagedResult<T>
+    {
+        public List<T> Items { get; set; }
+        public int Page { get; set; }
+        public byte PageSize { get; set; }
+        public int TotalCount { get; set; }
+        //public int TotalPages { get; set; }
```
Comment (clangelov) [inline SportsGoods.Core/Models/PagedResult.cs]: Remove commented code

File: SportsGoods.Core/Models/PagedResult.cs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SportsGoods.Core.Models
+{
+    public class PagedResult<T>
+    {
+        public List<T> Items { get; set; }
```
Comment (clangelov) [inline SportsGoods.Core/Models/PagedResult.cs]: Address the build action warning

File: SportsGoods.Web/SportsGoods.Web.csproj [hunk 6]
```diff
@@ -15,6 +15,8 @@
     <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="8.0.3" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
     <PackageReference Include="NUnit3TestAdapter" Version="4.5.0" />
+    <PackageReference Include="System.Linq.Dynamic.Core" Version="1.3.10" />
```
Comment (clangelov) [inline SportsGoods.Web/SportsGoods.Web.csproj]: do we really need this package?

File: UnitTests/Tests/GetAllProductsQueryHandlerTest.cs [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+            Assert.That(result, Is.Not.Null);
+            Assert.That(result.Items.Count, Is.EqualTo(2));
+        }
+
+        [Test]
+        public async Task Handle_ReturnsCorrectPageBasedOnPageSize()
+        {
+            var pageNumber = 1;
+            byte[] pageSizes = { 10, 5, 20 };
```
Comment (clangelov) [inline UnitTests/Tests/GetAllProductsQueryHandlerTest.cs]: you can use something like 
https://docs.nunit.org/articles/nunit/technical-notes/usage/Parameterized-Tests.html

and 
https://dzone.com/articles/nunit-tutorial-parameterized-tests-with-examples

to make the test more readable

File: UnitTests/Tests/GetAllProductsQueryHandlerTest.cs [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+                new Product { Id = Guid.NewGuid(), Title = "Product 2", Price = 20 },
+            };
+
+            return new PagedResult<Product>
+            {
+                Items = products,
+                Page = 1,
+                PageSize = 10,
+                TotalCount = products.Count
+                //TotalPages = 1
```
Comment (clangelov) [inline UnitTests/Tests/GetAllProductsQueryHandlerTest.cs]: remove unused code

File: SportsGoods.Api/Program.cs (hunk 1 above)
Comment (LVarsaminov) [reply SportsGoods.Api/Program.cs]: Done

File: SportsGoods.Api/SportsGoods.Api.csproj (hunk 2 above)
Comment (LVarsaminov) [reply SportsGoods.Api/SportsGoods.Api.csproj]: Done

File: SportsGoods.Core/Models/PagedResult.cs (hunk 4 above)
Comment (LVarsaminov) [reply SportsGoods.Core/Models/PagedResult.cs]: Done

File: SportsGoods.Web/SportsGoods.Web.csproj (hunk 6 above)
Comment (LVarsaminov) [reply SportsGoods.Web/SportsGoods.Web.csproj]: Done

File: UnitTests/Tests/GetAllProductsQueryHandlerTest.cs (hunk 7 above)
Comment (LVarsaminov) [reply UnitTests/Tests/GetAllProductsQueryHandlerTest.cs]: Done

File: UnitTests/Tests/GetAllProductsQueryHandlerTest.cs (hunk 8 above)
Comment (LVarsaminov) [reply UnitTests/Tests/GetAllProductsQueryHandlerTest.cs]: Done

File: SportsGoods.Core/Models/PagedResult.cs (hunk 5 above)
Comment (LVarsaminov) [reply SportsGoods.Core/Models/PagedResult.cs]: Fixed

File: SportsGoods.App/QueryHandlers/GetAllProductsQueryHandler.cs (hunk 3 above)
Comment (LVarsaminov) [reply SportsGoods.App/QueryHandlers/GetAllProductsQueryHandler.cs]: Done

**Review by clangelov — APPROVED** [review]
it's fine

**Pull request merged**